lib/blocks.php

```diff
@@ -256,22 +257,44 @@ static function() {
 
 		// If the file exists, enqueue it.
 		if ( file_exists( gutenberg_dir_path() . $theme_style_path ) ) {
-
-			if ( file_exists( $stylesheet_path ) ) {
-				// If there is a main stylesheet for this block, append the theme styles to main styles.
-				wp_add_inline_style(
-					"wp-block-{$block_name}",
-					file_get_contents( gutenberg_dir_path() . $theme_style_path )
-				);
+			$theme_style_handle = "wp-block-{$block_name}-theme";
+			if ( wp_style_is( $theme_style_handle, 'registered' ) ) {
+				wp_deregister_style( $theme_style_handle );
+				if ( file_exists( $stylesheet_path ) ) {
+					// If there is a main stylesheet for this block, append the theme styles to main styles.
+					wp_register_style(
+						$theme_style_handle,
+						false,
+						array(),
+						$default_version
+					);
+				} else {
+					// If there is no main stylesheet for this block, register theme style.
+					wp_register_style(
+						$theme_style_handle,
+						gutenberg_url( $theme_style_path ),
+						array(),
+						$default_version
+					);
+					wp_style_add_data( $theme_style_handle, 'path', gutenberg_dir_path() . $theme_style_path );
+				}
 			} else {
-				// If there is no main stylesheet for this block, register theme style.
-				wp_register_style(
-					"wp-block-{$block_name}",
-					gutenberg_url( $theme_style_path ),
-					array(),
-					$default_version
-				);
-				wp_style_add_data( "wp-block-{$block_name}", 'path', gutenberg_dir_path() . $theme_style_path );
+				if ( file_exists( $stylesheet_path ) ) {
+					// If there is a main stylesheet for this block, append the theme styles to main styles.
+					wp_add_inline_style(
+						$block_style_handle,
+						file_get_contents( gutenberg_dir_path() . $theme_style_path )
+					);
+				} else {
+					// If there is no main stylesheet for this block, register theme style.
+					wp_register_style(
+						$block_style_handle,
+						gutenberg_url( $theme_style_path ),
+						array(),
+						$default_version
+					);
+					wp_style_add_data( $block_style_handle, 'path', gutenberg_dir_path() . $theme_style_path );
+				}
```

Comment:
I think the logic in the outer `if {} else {}` clauses is (or rather should be almost the same. I find the implementation here a bit confusing with all the duplicate code. I think the code can remain closer to what it was before (only one if else, no nesting of the clauses). And unless I'm missing something, we can simply have the following as its own standalone if clause before the if else:
```php
if ( wp_style_is( $theme_style_handle, 'registered' ) ) {
	wp_deregister_style( $theme_style_handle );
}
```